tests/test_iotedgedev.py

```diff
@@ -1,266 +1,385 @@
 import json
 import os
-import shutil
 import platform
-
 import pytest
-from click.testing import CliRunner
+import shutil
+import time
 
 from iotedgedev.compat import PY35
 from iotedgedev.connectionstring import (DeviceConnectionString,
                                          IoTHubConnectionString)
 from iotedgedev.envvars import EnvVars
 from iotedgedev.output import Output
 
-from .utility import assert_json_file_equal
-from .utility import get_platform_type
+from .utility import (assert_json_file_equal,
+                      get_platform_type,
+                      get_all_docker_images,
+                      get_all_docker_containers,
+                      remove_docker_container,
+                      remove_docker_image,
+                      get_docker_os_type,
+                      runner_invoke)
 
 pytestmark = pytest.mark.e2e
 
-root_dir = os.getcwd()
-tests_dir = os.path.join(root_dir, "tests")
 output = Output()
 envvars = EnvVars(output)
+
+root_dir = os.getcwd()
+tests_dir = os.path.join(root_dir, "tests")
+
 env_file_name = envvars.get_dotenv_file()
 env_file_path = envvars.get_dotenv_path(env_file_name)
 
-test_solution = "test_solution"
-test_solution_dir = os.path.join(tests_dir, test_solution)
 launch_json_file = os.path.join(tests_dir, "assets", "launch.json")
 launch_json_file_without_nodejs = os.path.join(tests_dir, "assets", "launch_without_nodejs.json")
 
+test_solution = "test_solution"
+test_solution_dir = os.path.join(tests_dir, test_solution)
+
 test_solution_shared_lib = "test_solution_shared_lib"
 test_solution_shared_lib_dir = os.path.join(tests_dir, "assets", test_solution_shared_lib)
 
+templates = ["c", "csharp", "java", "nodejs", "python", "csharpfunction"]
+
+
+def create_solution(template, custom_module_name=None):
+    if custom_module_name is None:
+        module_name = template + "module"
+    else:
+        module_name = custom_module_name
+    result = runner_invoke(['new', test_solution, '-m', module_name, '-t', template])
+
+    return result
+
+
+def add_module(template):
+    module_name = template + "module"
+    result = runner_invoke(["solution", "add", module_name, '--template', template])
+    return result
 
-@pytest.fixture(scope="function", autouse=True)
-def create_solution(request):
-    def clean():
-        os.chdir(root_dir)
-        shutil.rmtree(test_solution_dir, ignore_errors=True)
 
-    clean()
+def clean_folder(folder_path):
+    os.chdir(tests_dir)
+    time.sleep(5)
```

Comment:
What's the `sleep` here for?

Reply:
it will throw exception for code "getcwd()" so I have to sleep here.